pulser-core/pulser/backend/default_observables.py

```diff
@@ -0,0 +1,369 @@
+    """
+
+    def __init__(
+        self,
+        operator: Operator,
+        *,
+        evaluation_times: Sequence[float] | None = None,
+    ):
+        """Initializes the observable."""
+        super().__init__(evaluation_times=evaluation_times)
+        # TODO: Checks
+        self.operator = operator
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return f"expectation_{self.index}"
+
+    def apply(self, *, state: State, **kwargs: Any) -> Any:
+        """Calculates the observable to store in the Results."""
+        return self.operator.expect(state)
+
+
+class CorrelationMatrix(Observable):
+    """Stores the correlation matrix for the current state.
+
+    Args:
+        evaluation_times: The relative times at which to compute the
+            correlation matrix. If left as `None`, uses the
+            `default_evaluation_times` of the backend's `EmulationConfig`.
+        one_state: The eigenstate to measure the population of in the
+            correlation matrix. Can be left undefined if the state's
+            eigenstates form a known eigenbasis with a defined "one state".
+    """
+
+    def __init__(
+        self,
+        *,
+        evaluation_times: Sequence[float] | None = None,
+        one_state: Eigenstate | None = None,
+    ):
+        """Initializes the observable."""
+        super().__init__(evaluation_times=evaluation_times)
+        # TODO: Checks
+        self.one_state = one_state
+
+    def name(self) -> str:
+        """Returns the name of the observable."""
+        return "correlation_matrix"
+
+    @staticmethod
+    @functools.cache
+    def _get_number_operator(
+        qudit_ids: frozenset[int],
+        n_qudits: int,
+        eigenstates: Sequence[Eigenstate],
+        one_state: Eigenstate,
+        op_type: Type[OperatorType],
+    ) -> OperatorType:
+        n_op = {one_state * 2: 1.0}
+        return op_type.from_operator_repr(
+            eigenstates=eigenstates,
+            n_qudits=n_qudits,
+            operations=[(1.0, [(n_op, qudit_ids)])],
+        )
+
+    def apply(
+        self, *, state: State, hamiltonian: Operator, **kwargs: Any
+    ) -> list[list]:
+        """Calculates the observable to store in the Results."""
+        try:
+            return state.get_correlation_matrix(one_state=self.one_state)
+        except NotImplementedError:
+            pass
+
+        @functools.cache
+        def calc_expectation(qudit_ids: frozenset[int]) -> Any:
+            return self._get_number_operator(
+                qudit_ids,
+                state.n_qudits,
+                state.eigenstates,
+                self.one_state or state.infer_one_state(),
+                type(hamiltonian),
+            ).expect(state)
+
+        return [
+            [
+                calc_expectation(frozenset((i, j)))
+                for j in range(state.n_qudits)
+            ]
+            for i in range(state.n_qudits)
+        ]
```

Comment:
We have recently had to add the following to this class

```
if hasattr(state, "get_correlation_matrix") and callable(
            state.get_correlation_matrix
        ):
            return state.get_correlation_matrix(
```

to resolve a performance bottleneck. We managed to improve the runtime of this callback a factor of 30 for the case we benchmarked. I would very much like to have some mechanism that prevents us from having to write an emu-mps specific callback to avoid a regression.

Reply:
I've done some thinking, and in the worst case, we can always override the Callback behaviour in the backend ourselves. This holds for all the comments I wrote in this file, so I don't consider any of it truly blocking, but some thought might deliver a more elegant solution.

Reply:
I think in the end it's best if you just take implementations that work well for state-vector, and we monkeypatch them in the emulators as needed.

Reply:
So, in your `MPSConfig` or `MPSConfig`, you would basically override `apply()` of select observables at runtime?